llvm/lib/Target/AIE/AIECombinerHelper.cpp

```diff
@@ -87,12 +87,43 @@ bool llvm::canDelayMemOp(MachineInstr &MemI, MachineInstr &Dest,
   auto InstrRange = make_range(MII, MIE);
   bool SawStore = MemI.mayStore();
   auto UnsafeToMovePast = [&](const MachineInstr &MI) {
-    return (isUseOf(MI, MemI) && !isTriviallyDead(MI, MRI)) ||
+    bool IsCopyBetween = false;
+    // We can delay an instruction after a copy, if the copy just
+    // connects MemI and Dest. After combining, this copy will be dead.
+    if (MI.isCopy() && MRI.hasOneNonDBGUse(MI.getOperand(1).getReg()) &&
+        MRI.hasOneNonDBGUse(MI.getOperand(0).getReg())) {
+      MachineInstr *CopyOrignMI = MRI.getVRegDef(MI.getOperand(1).getReg());
+      MachineInstr *CopyDestMI =
+          &*MRI.use_instr_nodbg_begin(MI.getOperand(0).getReg());
+      if (CopyOrignMI == &MemI && CopyDestMI == &Dest) {
+        IsCopyBetween = true;
+      }
+    }
+    return ((isUseOf(MI, MemI) && !IsCopyBetween) &&
+            !isTriviallyDead(MI, MRI)) ||
            !MI.isSafeToMove(nullptr, SawStore);
   };
   return none_of(InstrRange, UnsafeToMovePast);
 }
 
+/// Find the def instruction for \p Reg, folding away any trivial copies and
+/// bitcasts. May return nullptr if \p Reg is not a generic virtual register.
+MachineInstr *
+llvm::getDefIgnoringCopiesAndBitcasts(Register Reg,
+                                      const MachineRegisterInfo &MRI) {
+
+  MachineInstr *DefInstr = getDefIgnoringCopies(Reg, MRI);
+
+  while (
+      DefInstr && DefInstr->getOpcode() == TargetOpcode::G_BITCAST &&
+      MRI.hasOneNonDBGUse(
+          DefInstr->getOperand(0).getReg()) /*No other use for this bitcast*/) {
```

Comment:
Super-nit: I would take the `/*No other use for this bitcast*/` comment out of the loop condition to make it easier to read.